test/DotNetty.Handlers.Tests/SniHandlerTest.cs

```diff
@@ -0,0 +1,296 @@
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Linq;
+    using System.Net.Security;
+    using System.Security.Authentication;
+    using System.Security.Cryptography.X509Certificates;
+    using System.Threading.Tasks;
+    using DotNetty.Buffers;
+    using DotNetty.Common.Concurrency;
+    using DotNetty.Handlers.Tls;
+    using DotNetty.Tests.Common;
+    using DotNetty.Transport.Channels;
+    using DotNetty.Transport.Channels.Embedded;
+    using Xunit;
+    using Xunit.Abstractions;
+
+    public class SniHandlerTest : TestBase
+    {
+        static readonly TimeSpan TestTimeout = TimeSpan.FromSeconds(10);
+        static readonly Dictionary<string, X509Certificate2> CertMap = new Dictionary<string, X509Certificate2>();
+
+        static SniHandlerTest()
+        {
+            X509Certificate2 tlsCertificate = TestResourceHelper.GetTestCertificate();
+            X509Certificate2 tlsCertificate2 = TestResourceHelper.GetTestCertificate2();
+
+            CertMap[tlsCertificate.GetNameInfo(X509NameType.DnsName, false)] = tlsCertificate;
+            CertMap[tlsCertificate2.GetNameInfo(X509NameType.DnsName, false)] = tlsCertificate2;
+        }
+
+        public SniHandlerTest(ITestOutputHelper output)
+            : base(output)
+        {
+        }
+
+        public static IEnumerable<object[]> GetTlsReadTestData()
+        {
+            var random = new Random(Environment.TickCount);
+            var lengthVariations =
+                new[]
+                {
+                    new[] { 1 },
+                    new[] { 2, 8000, 300 },
+                    new[] { 100, 0, 1000 },
+                    new[] { 4 * 1024 - 10, 1, 0, 1 },
+                    new[] { 0, 24000, 0, 1000 },
+                    new[] { 0, 4000, 0 },
+                    new[] { 16 * 1024 - 100 },
+                    Enumerable.Repeat(0, 30).Select(_ => random.Next(0, 17000)).ToArray()
+                };
+            var boolToggle = new[] { false, true };
+            var protocols = new[] { SslProtocols.Tls12 };
+            var writeStrategyFactories = new Func<IWriteStrategy>[]
+            {
+                () => new AsIsWriteStrategy(),
+                () => new BatchingWriteStrategy(1, TimeSpan.FromMilliseconds(20), true),
+                () => new BatchingWriteStrategy(4096, TimeSpan.FromMilliseconds(20), true),
+                () => new BatchingWriteStrategy(32 * 1024, TimeSpan.FromMilliseconds(20), false)
+            };
+
+            return
+                from frameLengths in lengthVariations
+                from isClient in boolToggle
+                from writeStrategyFactory in writeStrategyFactories
+                from protocol in protocols
+                from targetHost in CertMap.Keys
+                select new object[] { frameLengths, isClient, writeStrategyFactory(), protocol, targetHost };
+        }
+
+
+        [Theory]
+        [MemberData(nameof(GetTlsReadTestData))]
```

Comment:
@nayato done. But haven't changed MemberData to InlineData since I am still feeding test data with multiple host names, so easier to do via MemberData. Just reduced the other inputs to 1 set in each.